zio-http/shared/src/main/scala/zio/http/codec/BinaryCodecWithSchema.scala

```diff
@@ -3,9 +3,21 @@ package zio.http.codec
 import zio.schema.Schema
 import zio.schema.codec.BinaryCodec
 
-final case class BinaryCodecWithSchema[A](codec: BinaryCodec[A], schema: Schema[A])
+trait CodecBuilder {
+  def build[A](schema: Schema[A]): BinaryCodec[A]
+}
+
+final case class BinaryCodecWithSchema[A](codecBuilder: CodecBuilder, schema: Schema[A]) {
```

Comment:
I don't like the structure of this:

1. `CodecBuilder` says, "I can give you a binary codec for any schema"
2. `BinaryCodecWithSchema` says "Here is a schema and a way to get a binary codec for ANY schema"

Why was this change necessary?